Python/import.c

```diff
@@ -1744,10 +1751,461 @@ import_find_and_load(PyThreadState *tstate, PyObject *abs_name)
+{
+    assert(module != NULL);
+    PyObject *lazy_submodules = PyObject_GetAttr(module, &_Py_ID(__lazy_submodules__));
+    if (lazy_submodules == NULL) {
+        PyErr_Clear();
+        return 0;
+    }
+    int res = PySet_Contains(lazy_submodules, name);
+    Py_DECREF(lazy_submodules);
+    if (res < 0) {
+        return -1;
+    }
+    return res;
+}
+
+static int
+lazy_loaded_add_parent(PyObject *module, PyObject *name)
+{
+    assert(module != NULL);
+    PyObject *lazy_submodules = PyObject_GetAttr(module, &_Py_ID(__lazy_submodules__));
+    if (lazy_submodules == NULL) {
+        PyErr_Clear();
+        lazy_submodules = PySet_New(NULL);
+        if (lazy_submodules == NULL) {
+            return -1;
+        }
+        if (PyObject_SetAttr(module, &_Py_ID(__lazy_submodules__), lazy_submodules) < 0) {
+            Py_DECREF(lazy_submodules);
+            return -1;
+        }
+    }
+    int res = PySet_Add(lazy_submodules, name);
+    Py_DECREF(lazy_submodules);
+    if (res < 0) {
+        return -1;
+    }
+    return 0;
+}
+
+static PyLazyImportObject *
+new_lazy_import(PyObject *parent, PyObject *child, PyObject *globals, PyObject *locals)
+{
+    PyObject *fromlist = PyTuple_New(1);
+    if (fromlist == NULL) {
+        return NULL;
+    }
+    Py_INCREF(child);
+    if (PyTuple_SetItem(fromlist, 0, child) < 0) {
+        return NULL;
+    }
+    PyObject *from = _PyLazyImport_NewModule(parent, globals, locals, fromlist, NULL);
+    Py_DECREF(fromlist);
+    if (from == NULL) {
+        return NULL;
+    }
+    PyLazyImportObject *lazy_import = (PyLazyImportObject *)_PyLazyImport_NewObject(from, child);
+    Py_DECREF(from);
+    return lazy_import;
+}
+
+static int
+feed_lazy_loaded(PyThreadState *tstate, PyObject *name)
+{
+    int ret = 1;
+    PyObject *lazy_attrubutes = tstate->interp->lazy_attrubutes;
+    if (lazy_attrubutes == NULL) {
+        lazy_attrubutes = PyDict_New();
+        if (lazy_attrubutes == NULL) {
+            return -1;
+        }
+        tstate->interp->lazy_attrubutes = lazy_attrubutes;
+    }
+    Py_INCREF(name);
+    while (true) {
+        Py_ssize_t dot = PyUnicode_FindChar(name, '.', 0, PyUnicode_GET_LENGTH(name), -1);
+        if (dot < 0) {
+            break;
+        }
+        PyObject *parent = PyUnicode_Substring(name, 0, dot);
+        if (parent == NULL) {
+            Py_DECREF(name);
+            return -1;
+        }
+        PyObject *child = PyUnicode_Substring(name, dot + 1, PyUnicode_GET_LENGTH(name));
+        if (child == NULL) {
+            Py_DECREF(parent);
+            Py_DECREF(name);
+            return -1;
+        }
+        PyObject *lazy_loaded_set = PyDict_GetItemWithError(lazy_attrubutes, parent);
+        if (lazy_loaded_set == NULL) {
+            if (PyErr_Occurred()) {
+                Py_DECREF(child);
+                Py_DECREF(parent);
+                Py_DECREF(name);
+                return -1;
```

Comment:
seems like an `error:` label that decrefs these three things and returns -1 could reduce a lot of noise in this function. maybe even a few different stacked error labels that decref different things

Reply:
That's what I initially had... I was in the borderline between the two options 😛

I'll see if I can improve it.